flax/experimental/nnx/nnx/nn/lora.py

```diff
@@ -0,0 +1,192 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import typing as tp
+
+import jax
+import jax.numpy as jnp
+
+from flax.experimental import nnx
+from flax.experimental.nnx.nnx import rnglib, variables
+from flax.experimental.nnx.nnx.module import Module
+from flax.experimental.nnx.nnx.nn import initializers
+from flax.typing import (
+  Dtype,
+  Initializer,
+)
+
+Array = jax.Array
+Axis = int
+Size = int
+A = tp.TypeVar('A')
+
+default_kernel_init = initializers.lecun_normal()
+
+
+class LoRAParam(variables.Variable[A]): pass
+
+
+class LoRA(Module):
+  """A standalone LoRA layer.
+
+  Example usage::
+
+    >>> from flax.experimental import nnx
+    >>> import jax, jax.numpy as jnp
+    >>> layer = nnx.LoRA(3, 4, 2, rngs=nnx.Rngs(0))
+    >>> layer.lora_a.value.shape
+    (3, 2)
+    >>> layer.lora_b.value.shape
+    (2, 4)
+    >>> # Wrap around existing layer
+    >>> linear = nnx.Linear(3, 4, rngs=nnx.Rngs(0))
+    >>> wrapper = nnx.LoRA(3, 4, 2, base_module=linear, rngs=nnx.Rngs(1))
+    >>> assert wrapper.base_module == linear
+    >>> wrapper.lora_a.value.shape
+    (3, 2)
+    >>> layer.lora_b.value.shape
+    (2, 4)
+    >>> y = layer(jnp.ones((16, 3)))
+    >>> y.shape
+    (16, 4)
+
+  Attributes:
+    in_features: the number of input features.
+    out_features: the number of output features.
+    lora_rank: the rank of the LoRA dimension.
+    base_module: a base module to call and substitute, if possible.
+    dtype: the dtype of the computation (default: infer from input and params).
+    param_dtype: the dtype passed to parameter initializers (default: float32).
+    precision: numerical precision of the computation see `jax.lax.Precision`
+      for details.
+    kernel_init: initializer function for the weight matrices.
+    use_lora_param_type: if yes, LoRA params will be of different Param type.
+      This makes it easier to split model state for trainable/nontrainables.
+  """
+  def __init__(
+    self,
+    in_features: int,
+    out_features: int,
+    lora_rank: int,
+    *,
+    base_module: tp.Optional[nnx.Module] = None,
+    dtype: tp.Optional[Dtype] = None,
+    param_dtype: Dtype = jnp.float32,
+    kernel_init: Initializer = default_kernel_init,
+    lora_param_type: bool = True,
+    rngs: rnglib.Rngs,
+  ):
+    self.in_features = in_features
+    self.out_features = out_features
+    self.dtype = dtype
+    self.param_dtype = param_dtype
+    self.lora_param_type = lora_param_type
+    self.base_module = base_module
+
+    param_type = LoRAParam if self.lora_param_type else nnx.Param
+    self.lora_a = param_type(
+      kernel_init(rngs.params(), (in_features, lora_rank), param_dtype)
+    )
+    self.lora_b = param_type(
```

Comment:
I'm wondering if there are more informative variable names we can use like `self.lora_in` and `self.lora_out`, or is `a` and `b` standard naming conventions for LoRA?

Reply:
Yeah unfortunately AFAIK `a` and `b` are the conventions here...